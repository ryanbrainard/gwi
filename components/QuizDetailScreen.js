import _ from "lodash";
import React from "react";
import { View } from "react-native";
import CharacterQuizCarousel from "./CharacterQuizCarousel";

export default class QuizDetailScreen extends React.Component {
  // TODO: how to do prop types on navigation params

  static navigationOptions = ({ navigation }) => ({
    title: navigation.getParam("charSet").name
  });

  render() {
    const { navigation } = this.props;
    const charSet = navigation.getParam("charSet");

    return (
      <View
        style={{
          flex: 1,
          alignItems: "center",
          justifyContent: "center"
        }}
      >
        <CharacterQuizCarousel characters={_.shuffle(charSet.characters)} />
      </View>
    );
  }
}
