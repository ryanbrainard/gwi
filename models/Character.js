import VoicePlayerMRU from "./VoicePlayerMRU";

export default class Character {
  static _chars = {};

  static register(char) {
    this._chars[char.name] = char;
    return char;
  }

  static find(name) {
    return this._chars[name];
  }

  constructor(name, group, voices) {
    this._name = name;
    this._group = group;
    this._voices = voices;

    this.player = new VoicePlayerMRU(this._voices);
  }

  get key() {
    return this.name;
  }

  get name() {
    return this._name;
  }

  get group() {
    return this._group.map(c => Character.find(c));
  }

  async preloadPlay() {
    await this.player.load();
  }

  async play() {
    await this.player.play();
  }
}
