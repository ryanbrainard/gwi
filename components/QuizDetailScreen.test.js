import React from "react";
import renderer from "react-test-renderer";
import Character from "../models/Character";
import QuizDetailScreen from "./QuizDetailScreen";

test("renders correctly", () => {
  jest.mock("Animated");

  const group = ["A", "B"];
  const characters = group.map(char =>
    Character.register(new Character(char, group))
  );

  const navigation = {
    getParam: jest.fn(() => ({ characters }))
  };

  let idIndex = 0;
  const idMaker = () => idIndex++;

  const elem = renderer.create(
    <QuizDetailScreen
      navigation={navigation}
      _shuffle={s => s}
      _genKey={idMaker}
    />
  );
  const inst = elem.getInstance();
  inst.setState({ layout: { width: 100, height: 200 } });

  expect(Object.keys(inst.state.items)).toEqual(["0", "1"]);
  expect(elem.toJSON()).toMatchSnapshot();

  inst.state.items["0"].answered = Character.find("B"); // incorrect
  expect(Object.keys(inst.state.items)).toEqual(["0", "1", "2"]); // added to end
  expect(inst.state.items["0"].char).toEqual(inst.state.items["2"].char); // same char
  expect(elem.toJSON()).toMatchSnapshot();

  inst.state.items["1"].answered = Character.find("B"); // correct
  expect(Object.keys(inst.state.items)).toEqual(["0", "1", "2"]); // not added to end
  expect(elem.toJSON()).toMatchSnapshot();

  inst.restartQuiz();
  expect(Object.keys(inst.state.items)).toEqual(["3", "4"]);
  expect(elem.toJSON()).toMatchSnapshot();
});
