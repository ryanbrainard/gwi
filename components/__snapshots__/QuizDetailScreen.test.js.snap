// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "height": 1000.5,
      }
    }
  >
    <RCTScrollView
      activeAnimationOptions={null}
      activeAnimationType="timing"
      activeSlideAlignment="center"
      activeSlideOffset={20}
      apparitionDelay={0}
      automaticallyAdjustContentInsets={false}
      autoplay={false}
      autoplayDelay={5000}
      autoplayInterval={3000}
      callbackOffsetMargin={5}
      collapsable={undefined}
      containerCustomStyle={Object {}}
      contentContainerCustomStyle={Object {}}
      contentContainerStyle={
        Array [
          Object {},
          Object {
            "paddingLeft": 37.5,
            "paddingRight": 37.5,
          },
        ]
      }
      data={Array []}
      decelerationRate="fast"
      directionalLockEnabled={true}
      disableVirtualization={false}
      enableMomentum={false}
      enableSnap={true}
      firstItem={0}
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={undefined}
      hasParallaxImages={false}
      horizontal={true}
      inactiveSlideOpacity={0.7}
      inactiveSlideScale={0.9}
      inactiveSlideShift={0}
      initialNumToRender={7}
      initialScrollIndex={undefined}
      invertStickyHeaders={false}
      inverted={false}
      itemWidth={675}
      keyExtractor={[Function]}
      layout="default"
      lockScrollTimeoutDuration={1000}
      lockScrollWhileSnapping={false}
      loop={true}
      loopClonesPerSide={3}
      maxToRenderPerBatch={15}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onResponderRelease={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      onSnapToItem={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onTouchStart={[Function]}
      overScrollMode="never"
      pinchGestureEnabled={false}
      removeClippedSubviews={true}
      renderItem={[Function]}
      scrollEventThrottle={1}
      scrollsToTop={false}
      shouldOptimizeUpdates={true}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      slideStyle={Object {}}
      sliderWidth={750}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "flexDirection": "row",
          "opacity": 0,
          "width": 750,
        }
      }
      swipeThreshold={20}
      updateCellsBatchingPeriod={50}
      useScrollView={false}
      vertical={false}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={15}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
